### ISSUE TITLE
Code folding

### ISSUE THREAD

**Issue by gggustafson** [dart-lang/sdk#34795]
Visual Studio Code Docs contains a section on code folding (https://code.visualstudio.com/docs/editor/codebasics#_folding). However Dart is not present in the list. Why not?

**Comment by bwilkerson** [comment]
@DanTup

**Comment by DanTup** [comment]
@gggustafson I'm guessing the list you're referring to is the one under:

> Regions can also be defined by markers defined by each language. The following languages currently have markers defined:

?

If so, that's because Dart doesn't currently have any standard syntax for regions (at least, as far as I know) so we don't generate any folding sections for this.

@bwilkerson Do you know if this has been discussed? I could only find https://github.com/dart-lang/sdk/issues/33938 which was closed by the author. I think to do this well, we'd need to have a concept of this regions in the AST to ensure the region starts/ends in a good place (eg. can't intersect another folding region).

**Comment by bwilkerson** [comment]
I don't know what that quote mean by "regions". Server has a notification to support folding ('analysis.folding'), but I don't know whether that matches with what VS Code needs. The `FoldingRegion`s that server returns cannot overlap, but can be nested (methods can be folded, as can classes).

**Comment by gggustafson** [comment]
@DanTup Yes that is the list. 

I'll be honest, I do not think code folding aids in either readability or maintainability. Until I started coding in flutter/dart I had no need for it. But to prepare code for StackOverflow review and comment, I had to write monolithic blocks of Dart code. Usually I would have broken each class out into its own **.dart** file. But that flies in the face of providing the SO community with a complete example. Code folding would be useful to eliminate portions of code that are not applicable to the problem.

FYI my rationale is discussed in Minimalist Coding Guidelines at https://www.codeproject.com/Articles/236430/Minimalist-Coding-Guidelines.

**Comment by DanTup** [comment]
If I'm understanding correctly, the request is for some standard syntax for marking an arbitrary start/end of some code, and having that collapse. For example, in C# you can use `#region` and `#endregion` and the editor will allow them to be collapsed ([see example here](https://i.imgur.com/XYgVXF3.png)). Many people consider them bad, though as is often the case, I think they had their uses (for example if your class had to implement some interface of many methods, being able to group them together and collapse them was useful).

I think there are a few possibilities:

1. It's a supported language construct in Dart
2. It's just embedded in comments and just understood by the analyzer to generate folding regions
3. It's embedded in comments and regex'ed out by the editor

I think 1 would be an expensive way to deliver it and I think making 3 reliable could be tricky (we have no AST and need to avoid overlapping regions).

It'd be good to have an idea of how much demand there is for something like this - maybe collecting 👍's on this issue would be useful indicator?

**Comment by gggustafson** [comment]
If this issue is not just an oversight (i.e., the editor missed something that Dart provided), and it would be expensive to implement the feature, I'd just suggest not to do it. I can certainly live with no folding.

**Comment by DanTup** [comment]
Yeah, it's not an oversight; it's definitely something we don't support right now (either in the server or the Dart plugin).

I'll leave it to @bwilkerson to device whether it's something that should be considered and/or whether this issue should remain open for people to add 👍's to or not. I'll rename the issue at least so it's easier for others to find if they do come looking for the same thing.

**Comment by jwren** [comment]
Assigning to @DanTup to track.

CC @devoncarew

**Comment by devoncarew** [comment]
Sounds like this is a feature request for the analysis server.

**Comment by tapizquent** [comment]
Have there been any updates on this? Will this be something that might be added?

**Comment by DanTup** [comment]
VS Code does now have the ability to mark arbitrary regions as being foldable:

https://code.visualstudio.com/updates/v1_70#_fold-selection

It's not quite what is being asked here (permanent markers in the Dart code that generate folding regions), but it might be useful for some cases nonetheless.
